File: src/main/java/weld/WeldValue.java
package weld;

import weld.types.StructType;
import weld.types.Type;

import java.nio.ByteBuffer;

import static weld.ByteBufferUtils.*;
import static weld.types.StructType.structOf;

/**
 * This is a wrapper around the weld value object.
 *
 * Note that this object must be closed after usage.
 */
public class WeldValue extends WeldManaged {
  /**
   * Size vecOf the value in bytes. -1 if the size is unknown.
   */
  private final long size;

  /**
   * Create a empty weld value.
   */
  public WeldValue() {
    this(allocateDirect(0));
  }

  /**
   * Create a weld value from a byte buffer.
   */
  public WeldValue(ByteBuffer buffer) {
    this(WeldJNI.weld_get_buffer_pointer(toDirect(buffer)), buffer.limit());
  }

  /**
   * Create a weld value that is pointing to a memory block allocated using
   * sun.misc.Unsafe or native code.
   */
  public WeldValue(long pointer) {
    this(pointer, -1L);
  }

  /**
   * Create a weld value that pointing to a memory block allocated using
   * sun.misc.Unsafe or native code.
   */
  public WeldValue(long pointer, long size) {
    this(WeldJNI.weld_value_new(pointer), size, false);
  }

  WeldValue(long handle, long size, boolean dummy) {
    super(handle);
    if (size < -1L) {
      throw new IllegalArgumentException("Size (" + size + ") must be >= 0, or -1 (unknown).");
    }
    this.size = size;
  }

  /**
   * Free the weld value.
   */
  @Override
  protected void doClose() {
    WeldJNI.weld_value_free(this.handle);
  }

  /**
   * Get the size vecOf the weld value in bytes.
   */
  public long size() {
    checkAccess();
    return this.size;
  }

  /**
   * Get the address to the data this value encapsulates.
   */
  public long getPointer() {
    checkAccess();
    return WeldJNI.weld_value_pointer(this.handle);
  }

  /**
   * Get the weld run ID this value is associated with.
   */
  public long getRunId() {
    checkAccess();
    return WeldJNI.weld_value_run(this.handle);
  }

  /**
   * Get the result vecOf a weld module run.
   */
  public WeldStruct result(StructType type) {
    checkAccess();
    return new WeldStruct(getPointer(), type);
  }

  /**
   * Get the result vecOf a weld module run.
   */
  public WeldStruct result(Type... types) {
    checkAccess();
    return new WeldStruct(getPointer(), structOf(types));
  }
}

File: src/main/scala/weld/WeldValue.scala
package weld

import java.nio.{ByteBuffer, ByteOrder}

import scala.annotation.varargs

/**
 * This is a wrapper around the weld value object.
 *
 * Note that this object must be closed after usage.
 */
class WeldValue private[weld](handle: Long, val size: Long = -1) extends WeldManaged(handle) {
  override protected def doClose(): Unit = WeldJNI.weld_value_free(handle)

  /**
   * Get the address to the data this value encapsulates.
   */
  def getPointer: Long = {
    checkAccess()
    WeldJNI.weld_value_pointer(handle)
  }

  /**
   * Get the weld run ID this value is associated with.
   */
  def getRunId: Long = {
    checkAccess()
    WeldJNI.weld_value_run(handle)
  }

  /**
   * Get the result of a weld module run.
   */
  def result(structType: StructType): WeldStruct = {
    checkAccess()
    new WeldStruct(getPointer, structType)
  }

  /**
   * Get the result of a weld module run.
   */
  @varargs
  def result(fieldTypes: WeldType*): WeldStruct = result(StructType(fieldTypes))
}

object WeldValue {
  private def directByteBuffer(size: Int) = {
    ByteBuffer.allocateDirect(size).order(ByteOrder.nativeOrder)
  }

  /**
   * Create an empty [[WeldValue]]
   */
  def empty(): WeldValue = apply(directByteBuffer(0))

  /**
   * Create a [[WeldValue]] using a (direct) [[ByteBuffer]].
   */
  def apply(buffer: ByteBuffer): WeldValue = {
    val direct = if (!buffer.isDirect) {
      val replacement = directByteBuffer(buffer.limit())
      replacement.put(buffer)
    } else {
      buffer
    }
    apply(WeldJNI.weld_get_buffer_pointer(direct), buffer.limit())
  }

  /**
   * Create a [[WeldValue]] from the given pointer.
   */
  def apply(pointer: Long): WeldValue = apply(pointer, -1L)

  /**
   * Create a sized [[WeldValue]] from the given pointer.
   */
  def apply(pointer: Long, size: Long): WeldValue = {
    new WeldValue(WeldJNI.weld_value_new(pointer), size)
  }
}
